content/recipes/swagger.md

```diff
@@ -77,75 +73,179 @@ async create(@Body() createCatDto: CreateCatDto) {
+```
+
+<figure><img src="/assets/enum_query.gif" /></figure>
+
+With `isArray` set to **true**, the `enum` can be selected as a **multi-select**:
+
+<figure><img src="/assets/enum_query_array.gif" /></figure>
+
+#### Arrays
+
+We have to manually indicate a type when the property is an array:
+
+```typescript
+@ApiProperty({ type: [String] })
+names: string[];
+```
+
+Simply put your type as the first element of an array (as shown above) or set an `isArray` property to `true`.
+
+#### Circular dependencies
+
+When you have cross-references between your classes, ensure to use lazy functions in order to provide a module with a type information lazily.
+
+```typescript
+@ApiProperty({ type: () => Node })
+node: Node;
+```
+
+#### Generics and interfaces
+
+TypeScript does not store any metadata about generics and interfaces. Hence, if you use them to define your DTOs, we may not be able to properly generate model definitons at the runtime.
+
+#### Raw definitions
+
+In some specific scenarios (e.g. deeply nested arrays, matrices), you may want to describe your type by hand.
+
+```typescript
+@ApiProperty({
+  type: 'array',
+  items: {
+    type: 'array',
+    items: {
+      type: 'number',
+    },
+  },
+})
+coords: number[][];
+```
+
+Likewise, in order to define your input/output content manually in controller classes, use `schema` property:
+
+```typescript
+@ApiBody({
+  schema: {
+    type: 'array',
+    items: {
+      type: 'array',
+      items: {
+        type: 'number',
+      },
+    },
+  },
+})
+async create(@Body() coords: number[][]) {}
+```
+
+#### Extra models
+
+In order to define additional models that should be inspected by Swagger module, use `@ApiExtraModels()` decorator:
+
+```typescript
+@ApiExtraModels(ExtraModel)
+export class CreateCatDto {}
+```
+
+#### oneOf, anyOf, allOf
+
+In order to combine schemas, you can use `oneOf`, `anyOf` or `allOf` keywords ([read more](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/)).
+
+```typescript
+@ApiProperty({
+  oneOf: [
+    { $ref: getSchemaPath(Cat) },
+    { $ref: getSchemaPath(Dog) },
+  ],
+})
+pet: Cat | Dog;
+```
+
+> info **Hint** `getSchemaPath()` function is imported from `@nestjs/swagger`.
+
+Both `Cat` and `Dog` has to be defined as extra models using `@ApiExtraModels()` decorator (at the class-level).
 
 #### Multiple specifications
 
-Swagger module also provides a way to support multiple specifications. In other words, you can serve different documentations with different `SwaggerUI` on different endpoints.
+Swagger module provides a way to support multiple specifications. In other words, you can serve different documentations with different UIs on different endpoints.
 
 In order to allow `SwaggerModule` to support multi-spec, your application must be written with modular approach. The `createDocument()` method takes in a 3rd argument: `extraOptions` which is an object where a property `include` expects an array of modules.
```

Comment:
In order to allow `SwaggerModule` to support multi-spec, your application must be written with modular approach. The `createDocument()` method takes in a 3rd argument: `extraOptions` which is an object where a property `include` expects an array of modules.

->

To support multiple specifications, your application must be written with a modular approach. The `createDocument()` method takes in a 3rd argument, `extraOptions`, which is an object with a the property `include`.  The `include` property has a value which is an array of modules.